includes/modules/core/command/PrivmsgCommand.hpp

```diff
@@ -17,17 +17,18 @@ class PrivmsgCommand : public CommandExecutor {
 	void execute(const Command& cmd, CommandSender& sender) {
 
 		vector<CommandSender *>& target = cmd.getArg<vector<CommandSender *> >("msgtarget");
-		string& message = cmd.getArg<string>("message");
+		string message = cmd.getArg<string>("message") + "\n";
 
 		for (vector<CommandSender *>::iterator it = target.begin(); it != target.end(); it++) {
-			try {
-				Channel *channel = dynamic_cast<Channel *>(*it);
-				if (!channel->isOnChan(static_cast<User *>(&sender))) {
-					sender.send(ResponseTypes::ERR_NOTONCHANNEL((*it)->getName().c_str(), sender.getName().c_str()));
-					continue ;
+			if(Channel::isValidIdentifier((*it)->getName())) {
+				if (Irc::getInstance().getChannels().find((*it)->getName()) != Irc::getInstance().getChannels().end()) {
+					if (!Irc::getInstance().getChannels().find((*it)->getName())->second->isOnChan(static_cast<User *>(&sender))) {
+						sender.send(ResponseTypes::ERR_NOTONCHANNEL((*it)->getName().c_str(), sender.getName().c_str()));
+						continue ;
+					}
 				}
-			} catch(std::bad_cast& e){}
-			(*it)->send(message.c_str());
+			}
+			(*it)->send(sender.getName() + " :" + message.c_str());
```

Comment:
To check with the RFC norme. Each response is supposed to be formatted as a command, so with a prefix and a command/response code.